File: workspace/ASTIC-lib/src/ro/info/asticlib/io/AsticStream.java
package ro.info.asticlib.io;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class AsticStream {
	public static final int BUFFER_SIZE = 1024;
	public InputStream in;
	public OutputStream out;
    private byte endSep = (byte)'|';
    private byte[] buffer;
    private boolean bufferDirty = false;

    public AsticStream(InputStream in,OutputStream out)
    {
        this.in = in;
        this.out = out;
        buffer = new byte [BUFFER_SIZE];
    }

    public String ReadString() throws IOException
    {
        if (bufferDirty)
        {
            clearBuffer();
        }
        StringBuilder builder  =  new StringBuilder();
        byte abyte = 0;
        do
        {
            int read = in.read(buffer, 0, buffer.length);
            bufferDirty = true;
            if(read > 0){
            	abyte = buffer[read - 1];
            	builder.append(new String(buffer,0,(abyte == endSep?read - 1:read),"US-ASCII"));
            }else{
            	abyte = endSep;
            }
        } while (abyte != endSep);
        return builder.toString();
    }

    public void WriteString(String message) throws IOException
    {
        byte [] messageBytes = message.getBytes();
        int messageSize = messageBytes.length;
        byte [] buffer = new byte[messageSize + 1];
        buffer[messageSize] = endSep;
        for (int i = buffer.length - 2; i >= 0; i-- )
        {
            buffer[i] = messageBytes[i];
        }
        out.write(buffer, 0, buffer.length);
    }

    private void clearBuffer()
    {
        for (int i = buffer.length - 1; i >= 0; i--)
        {
            buffer[i] = 0;
        }
        bufferDirty = false;
    }
}

File: ASTIC_Client_VS2010Ex/ASTIC_Client_VS2010Ex/ASTICStream.cs
using ASTIC_client;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ASTIC_client_V2
{
    public class ASTICStream
    {
        public const int BUFFER_SIZE = 1024;
        private Stream stream;
        private byte endSep = (byte)'|';
        private byte[] buffer;
        private bool bufferDirty = false;

        public ASTICStream(Stream stream)
        {
            this.stream = stream;
            buffer = new byte [BUFFER_SIZE];
        }

        public String ReadString()
        {
            if (bufferDirty)
            {
                clearBuffer();
            }
            StringBuilder builder  =  new StringBuilder();
            byte abyte = 0;
            do
            {
                int read = stream.Read(buffer, 0, buffer.Count());
                bufferDirty = true;
                abyte = buffer[read - 1];
                builder.Append(EncodeUtil.decode(buffer, abyte == endSep ? read - 1 : read));
            } while (abyte != endSep);
            return builder.ToString();
        }

        public void WriteString(String message)
        {
            int messageSize = System.Text.ASCIIEncoding.ASCII.GetByteCount(message);
            byte [] buffer = new byte[messageSize + 1];
            byte [] messageBytes = EncodeUtil.encode(message);
            buffer[messageSize] = endSep;
            for (int i = buffer.Count() - 2; i >= 0; i-- )
            {
                buffer[i] = messageBytes[i];
            }
            stream.Write(buffer, 0, buffer.Count());
        }

        private void clearBuffer()
        {
            for (int i = buffer.Count() - 1; i >= 0; i--)
            {
                buffer[i] = 0;
            }
            bufferDirty = false;
        }
    }
}
